### ISSUE TITLE
Caffeine-based cache (SyncCacheApi) loose thread locals when using a custom execution context

### ISSUE THREAD

**Issue by gokhanoner** [playframework/playframework#10957]
Play Version: 2.8.8

API: Java

Due to Caffeine based cache impl doesn't have a dedicated SyncCache implementation & uses DefaultSyncCache over async cache, see [here](https://github.com/playframework/playframework/blob/master/cache/play-caffeine-cache/src/main/scala/play/api/cache/caffeine/CaffeineCacheApi.scala#L155) , it loses thread locals, when using a custom execution context.

Solution is probably implement a proper NamedSyncCache implementation, similar to https://github.com/playframework/playframework/blob/master/cache/play-caffeine-cache/src/main/java/play/cache/caffeine/NamedCaffeineCache.java

**Comment by mkurz** [comment]
@gokhanoner Maybe I am confused, but after accessing the cache you should be back on the original thread anyway or not? Or do you want to use the ThreadLocal inside the cache operation?

**Comment by gokhanoner** [comment]
Yes. I need that ThreadLocal inside the cache operation. When using
ehcache, it works as expected, since its SyncCache is actually executed in
the same thread, rather than wrapped around the async one.
